staging/src/k8s.io/apimachinery/pkg/runtime/serializer/cbor/internal/modes/roundtrip_test.go

```diff
@@ -0,0 +1,304 @@
+			obj:  int8(math.MaxInt8),
+		},
+		{
+			name: "int8 min",
+			obj:  int8(math.MinInt8),
+		},
+		{
+			name: "int8 zero",
+			obj:  int8(math.MinInt8),
+		},
+		{
+			name: "uint8 max",
+			obj:  uint8(math.MaxUint8),
+		},
+		{
+			name: "uint8 zero",
+			obj:  uint8(0),
+		},
+		{
+			name: "float64",
+			obj:  float64(2.71),
+		},
+		{
+			name: "float64 max",
+			obj:  float64(math.MaxFloat64),
+		},
+		{
+			name: "float64 smallest nonzero",
+			obj:  float64(math.SmallestNonzeroFloat64),
+		},
+		{
+			name: "float64 no fractional component",
+			obj:  float64(5),
+		},
+		{
+			name: "float32",
+			obj:  float32(2.71),
+		},
+		{
+			name: "float32 max",
+			obj:  float32(math.MaxFloat32),
+		},
+		{
+			name: "float32 smallest nonzero",
+			obj:  float32(math.SmallestNonzeroFloat32),
+		},
+		{
+			name: "float32 no fractional component",
+			obj:  float32(5),
+		},
+		{
+			name: "time.Time",
+			obj:  time.Date(2222, time.May, 4, 12, 13, 14, 123, time.UTC),
+		},
+		{
+			name: "int64 omitempty",
+			obj: struct {
+				V int64 `json:"v,omitempty"`
+			}{},
+		},
+		{
+			name: "float64 omitempty",
+			obj: struct {
+				V float64 `json:"v,omitempty"`
+			}{},
+		},
+		{
+			name: "string omitempty",
+			obj: struct {
+				V string `json:"v,omitempty"`
+			}{},
+		},
+		{
+			name: "bool omitempty",
+			obj: struct {
+				V bool `json:"v,omitempty"`
+			}{},
+		},
+		{
+			name: "nil pointer omitempty",
+			obj: struct {
+				V *struct{} `json:"v,omitempty"`
+			}{},
+		},
+		{
+			name: "nil pointer to slice as struct field",
+			obj: struct {
+				V *[]interface{} `json:"v"`
+			}{},
+		},
+		{
+			name: "nil pointer to slice as struct field with omitempty",
+			obj: struct {
+				V *[]interface{} `json:"v,omitempty"`
+			}{},
+		},
+		{
+			name: "nil pointer to map as struct field",
+			obj: struct {
+				V *map[string]interface{} `json:"v"`
+			}{},
+		},
+		{
+			name: "nil pointer to map as struct field with omitempty",
+			obj: struct {
+				V *map[string]interface{} `json:"v,omitempty"`
+			}{},
+		},
+	} {
+		mps := tc.modePairs
+		if len(mps) == 0 {
+			// Default is all modes to all modes.
+			mps = []modePair{}
+			for _, em := range allEncModes {
+				for _, dm := range allDecModes {
+					mps = append(mps, modePair{enc: em, dec: dm})
+				}
+			}
+		}
+
+		for _, mp := range mps {
+			encModeName, ok := encModeNames[mp.enc]
+			if !ok {
+				t.Fatal("test case configured to run against unrecognized encode mode")
+			}
+
+			decModeName, ok := decModeNames[mp.dec]
+			if !ok {
+				t.Fatal("test case configured to run against unrecognized decode mode")
+			}
+
+			t.Run(fmt.Sprintf("enc=%s/dec=%s/%s", encModeName, decModeName, tc.name), func(t *testing.T) {
+				original := tc.obj
+
+				b, err := mp.enc.Marshal(original)
+				if err != nil {
+					t.Fatalf("unexpected error from Marshal: %v", err)
+				}
+
+				final := reflect.New(reflect.TypeOf(original))
+				err = mp.dec.Unmarshal(b, final.Interface())
```

Comment:
So this is testing decoding into a known type, but not decoding into a `map[string]interface{}`.  This is ok, but means it does not fully cover the "preserves distinction" tests